shared/domains/chelonia/GIMessage.js

```diff
@@ -50,61 +62,143 @@ export type GIOpKeyUpdate = {
+        if (key.meta?.private?.content) {
+          key.meta.private.content = encryptedIncomingData(contractID, state, key.meta.private.content, height, additionalKeys, headJSON, (value) => {
+          // Validator function to verify the key matches its expected ID
+            const computedKeyId = keyId(value)
+            if (computedKeyId !== key.id) {
+              throw new Error(`Key ID mismatch. Expected to decrypt key ID ${key.id} but got ${computedKeyId}`)
+            }
+          })
+        }
+        if (key.meta?.keyRequest?.contractID) {
+          try {
+            key.meta.keyRequest.contractID = maybeEncryptedIncomingData(contractID, state, key.meta.keyRequest.contractID, height, additionalKeys, headJSON)?.valueOf()
+          } catch {
+            delete key.meta.keyRequest.contractID
+          }
+        }
+      })
+    })
+  }
+
+  // If the operation is GIMessage.OP_CONTRACT,
+  // extract encrypted data from keys?.[].meta?.private?.content
+  if (op === GIMessage.OP_CONTRACT) {
+    (message: any).keys = (message: any).keys?.map((key, eKeyId) => {
+      return maybeEncryptedIncomingData(contractID, state, key, height, additionalKeys, headJSON, (key) => {
+        if (!key.meta?.private?.content) return
+        const decryptionFn = message.foreignContractID ? encryptedIncomingForeignData : encryptedIncomingData
+        // $FlowFixMe
+        const decryptionContract = ((message.foreignContractID ? message.foreignContractID : contractID): string)
+        key.meta.private.content = decryptionFn(decryptionContract, state, key.meta.private.content, height, additionalKeys, headJSON, (value) => {
           const computedKeyId = keyId(value)
           if (computedKeyId !== key.id) {
             throw new Error(`Key ID mismatch. Expected to decrypt key ID ${key.id} but got ${computedKeyId}`)
           }
         })
-      }
+      })
     })
   }
-  if ([GIMessage.OP_CONTRACT, GIMessage.OP_KEY_SHARE].includes(op)) {
-    (message: any).keys?.forEach((key) => {
-      if (key.meta?.private?.content) {
+
+  // If the operation is GIMessage.OP_KEY_SHARE,
+  // extract encrypted data from keys?.[].meta?.private?.content
+  if (op === GIMessage.OP_KEY_SHARE) {
+    return maybeEncryptedIncomingData(contractID, state, (message: any), height, additionalKeys, headJSON, (message) => {
+      (message: any).keys?.forEach((key) => {
+        if (!key.meta?.private?.content) return
         const decryptionFn = message.foreignContractID ? encryptedIncomingForeignData : encryptedIncomingData
-        const decryptionContract = message.foreignContractID ? message.foreignContractID : contractID
-        key.meta.private.content = decryptionFn(decryptionContract, state, key.meta.private.content, height, additionalKeys, (value) => {
+        const decryptionContract = String(message.foreignContractID ? message.foreignContractID : contractID)
```

Comment:
Can't this be simplified to:

```js
message.foreignContractID || contractID
```

without the `String` wrapper?

Reply:
The `String` wrapper is for casting as it needs to be a string.